numpy/lib/function_base.py

```diff
@@ -4756,55 +4813,83 @@ def _quantile(
+                                                          values_count)
+            # --- Sorting
+            arr.partition(
+                np.unique(np.concatenate(([0, -1],
+                                          previous_indexes.ravel(),
+                                          next_indexes.ravel(),
+                                          ))),
+                axis=0)
+            if supports_nans:
+                slices_having_nans = np.isnan(arr[-1, ...])
+            else:
+                slices_having_nans = None
+            # --- Get values from indexes
+            previous = arr[previous_indexes]
+            next = arr[next_indexes]
+            # --- Linear interpolation
+            gamma = _get_gamma(virtual_indexes, previous_indexes, method)
+            result_shape = virtual_indexes.shape + (1,) * (arr.ndim - 1)
+            gamma = gamma.reshape(result_shape)
+            result = _lerp(previous,
+                        next,
+                        gamma,
+                        out=out)
     else:
-        previous_indexes, next_indexes = _get_indexes(arr,
-                                                      virtual_indexes,
-                                                      values_count)
-        # --- Sorting
-        arr.partition(
-            np.unique(np.concatenate(([0, -1],
-                                      previous_indexes.ravel(),
-                                      next_indexes.ravel(),
-                                      ))),
-            axis=0)
-        if supports_nans:
-            slices_having_nans = np.isnan(arr[-1, ...])
-        else:
-            slices_having_nans = None
-        # --- Get values from indexes
-        previous = arr[previous_indexes]
-        next = arr[next_indexes]
-        # --- Linear interpolation
-        gamma = _get_gamma(virtual_indexes, previous_indexes, method)
-        result_shape = virtual_indexes.shape + (1,) * (arr.ndim - 1)
-        gamma = gamma.reshape(result_shape)
-        result = _lerp(previous,
-                       next,
-                       gamma,
-                       out=out)
+        # Weighted case, we need to sort anyway.
+        # This implements method="inverted_cdf".
+        weights = np.asanyarray(weights)
+        if axis != 0:
+            weights = np.moveaxis(weights, axis, destination=0)
+        index_array = np.argsort(arr, axis=axis, kind="stable")
+        # TODO: Which value to set to slices_having_nans.
+        slices_having_nans = None
+        # TODO: Deal with nan values, e.g. np.isnan(arr(index_array[-1])) by
+        # be true.
+        # TODO: Special case quantiles == 0 and quantiles == 1?
+        arr = np.take_along_axis(arr, index_array, axis=axis)
+        weights = np.take_along_axis(weights, index_array, axis=axis)
+        # Normalize weights to sum to one such that it corresponds to the
+        # empirical distribution function.
```

Comment:
Would be tempted to throw away weights here and keep only the cummulative sum (and use  `[-1]` for the normalization.